File: apps/ios/GalaxySSI/AgentNativeToolResourceLocks.swift
import CryptoKit
import Foundation

enum AgentNativeResourceLockMode: Equatable {
  case read
  case write
}

struct AgentNativeResourceLockRequest: Equatable {
  var key: String
  var mode: AgentNativeResourceLockMode
}

struct AgentNativeResourceLockPlan: Equatable {
  var requests: [AgentNativeResourceLockRequest]
  var resourceScoped: Bool

  func conflicts(with other: AgentNativeResourceLockPlan) -> Bool {
    let right = Dictionary(uniqueKeysWithValues: other.requests.map { ($0.key, $0.mode) })
    return requests.contains { left in
      guard let matching = right[left.key] else { return false }
      return left.mode == .write || matching == .write
    }
  }
}

enum AgentNativeToolResourcePolicy {
  static func resolve(
    descriptor: AgentNativeToolDescriptor,
    input: AgentMcpJSONObject,
    fallbackWorkspaceId: String = ""
  ) -> AgentNativeResourceLockPlan {
    let operationMode: AgentNativeResourceLockMode = descriptor.concurrency == .parallelReadOnly
      ? .read
      : .write
    if requiresExclusiveVerification(input) || exclusiveGlobalMutationToolIds.contains(descriptor.id) {
      return globalPlan(.write)
    }
    guard supportsWorkspaceScoping(descriptor) else {
      return globalPlan(operationMode)
    }
    let workspaceId = workspaceIdentity(
      descriptor: descriptor,
      input: input,
      fallbackWorkspaceId: fallbackWorkspaceId
    )
    guard !workspaceId.isEmpty else { return globalPlan(operationMode) }

    let workspaceSegment = stableSegment(workspaceId)
    let workspaceKey = "workspace:\(workspaceSegment)"
    let pathAccesses = descriptor.capabilities.contains(where: { $0.contains("workspace") })
      ? collectPathAccesses(descriptor: descriptor, input: input, operationMode: operationMode)
      : []
    var requests = [AgentNativeResourceLockRequest(key: globalKey, mode: .read)]
    if pathAccesses.isEmpty {
      requests.append(AgentNativeResourceLockRequest(key: workspaceKey, mode: operationMode))
    } else {
      requests.append(AgentNativeResourceLockRequest(key: workspaceKey, mode: .read))
      for access in pathAccesses {
        let components = normalizePath(access.path)
        if components.isEmpty {
          requests.append(AgentNativeResourceLockRequest(key: workspaceKey, mode: access.mode))
          continue
        }
        for index in components.indices {
          let prefix = components.prefix(index + 1).joined(separator: "/")
          requests.append(AgentNativeResourceLockRequest(
            key: "path:\(workspaceSegment):\(stableSegment(prefix))",
            mode: index == components.index(before: components.endIndex) ? access.mode : .read
          ))
        }
      }
    }
    return AgentNativeResourceLockPlan(requests: normalizeRequests(requests), resourceScoped: true)
  }

  static func resolveAction(
    descriptor: AgentNativeToolDescriptor,
    action: AgentAction,
    fallbackWorkspaceId: String = ""
  ) -> AgentNativeResourceLockPlan? {
    let raw = (action.parameters["input_json"] ?? "").ifBlank("{}")
    guard let data = raw.data(using: .utf8),
          let input = try? JSONDecoder().decode(AgentMcpJSONObject.self, from: data) else {
      return nil
    }
    return resolve(descriptor: descriptor, input: input, fallbackWorkspaceId: fallbackWorkspaceId)
  }

  private static func supportsWorkspaceScoping(_ descriptor: AgentNativeToolDescriptor) -> Bool {
    descriptor.capabilities.contains { capability in
      capability.contains("workspace") || capability.hasPrefix("runtime.") || capability.hasPrefix("project.")
    }
  }

  private static func workspaceIdentity(
    descriptor: AgentNativeToolDescriptor,
    input: AgentMcpJSONObject,
    fallbackWorkspaceId: String
  ) -> String {
    let explicit = clean(input["workspace_id"]?.stringValue ?? "", limit: maxIdentifierCharacters)
    if !explicit.isEmpty { return "\(descriptor.location.rawValue):\(explicit)" }
    let desktopId = clean(input["desktop_id"]?.stringValue ?? "", limit: maxIdentifierCharacters)
    if !desktopId.isEmpty {
      return "\(descriptor.location.rawValue):\(desktopId):\(fallbackWorkspaceId.ifBlank("default"))"
    }
    let fallback = clean(fallbackWorkspaceId, limit: maxIdentifierCharacters)
    return fallback.isEmpty ? "" : "\(descriptor.location.rawValue):\(fallback)"
  }

  private static func collectPathAccesses(
    descriptor: AgentNativeToolDescriptor,
    input: AgentMcpJSONObject,
    operationMode: AgentNativeResourceLockMode
  ) -> [PathAccess] {
    var accesses: [PathAccess] = []
    func visit(key: String, value: AgentMcpJSONValue) {
      switch value {
      case .string(let path) where isPathKey(key):
        accesses.append(PathAccess(
          path: path,
          mode: pathMode(toolId: descriptor.id, key: key, input: input, operationMode: operationMode)
        ))
      case .object(let object):
        object.forEach { visit(key: $0.key, value: $0.value) }
      case .array(let values):
        for value in values {
          if isPathCollectionKey(key), case .string(let path) = value {
            accesses.append(PathAccess(
              path: path,
              mode: pathMode(toolId: descriptor.id, key: key, input: input, operationMode: operationMode)
            ))
          } else {
            visit(key: key, value: value)
          }
        }
      default:
        break
      }
    }
    input.forEach { visit(key: $0.key, value: $0.value) }
    return accesses.filter { !$0.path.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
  }

  private static func pathMode(
    toolId: String,
    key: String,
    input: AgentMcpJSONObject,
    operationMode: AgentNativeResourceLockMode
  ) -> AgentNativeResourceLockMode {
    guard operationMode == .write else { return .read }
    let normalized = key.lowercased()
    if normalized == "source_path" || normalized == "source_paths" {
      return toolId.hasSuffix(".move") ? .write : .read
    }
    if normalized == "paths" { return .read }
    if normalized == "archive_path", toolId.hasSuffix(".zip.extract") || toolId.hasSuffix(".zip.list") {
      return .read
    }
    if normalized == "path", input["output_path"] != nil { return .read }
    return .write
  }

  private static func normalizePath(_ value: String) -> [String] {
    var normalized: [String] = []
    for raw in value.replacingOccurrences(of: "\\", with: "/").split(separator: "/", omittingEmptySubsequences: false) {
      let component = String(raw).trimmingCharacters(in: .whitespacesAndNewlines)
      switch component {
      case "", ".":
        continue
      case "..":
        if !normalized.isEmpty { normalized.removeLast() }
      default:
        normalized.append(String(component.prefix(maxPathComponentCharacters)))
      }
      if normalized.count == maxPathComponents { break }
    }
    return normalized
  }

  private static func normalizeRequests(
    _ requests: [AgentNativeResourceLockRequest]
  ) -> [AgentNativeResourceLockRequest] {
    Dictionary(grouping: requests, by: \.key).map { key, grouped in
      AgentNativeResourceLockRequest(
        key: key,
        mode: grouped.contains(where: { $0.mode == .write }) ? .write : .read
      )
    }.sorted { $0.key < $1.key }
  }

  private static func requiresExclusiveVerification(_ input: AgentMcpJSONObject) -> Bool {
    let verification = clean(input["verification_kind"]?.stringValue ?? "", limit: maxIdentifierCharacters)
      .lowercased()
    return !verification.isEmpty && verification != "none"
  }

  private static func globalPlan(_ mode: AgentNativeResourceLockMode) -> AgentNativeResourceLockPlan {
    AgentNativeResourceLockPlan(
      requests: [AgentNativeResourceLockRequest(key: globalKey, mode: mode)],
      resourceScoped: false
    )
  }

  private static func stableSegment(_ value: String) -> String {
    SHA256.hash(data: Data(value.utf8)).prefix(16).map { String(format: "%02x", $0) }.joined()
  }

  private static func clean(_ value: String, limit: Int) -> String {
    String(value.trimmingCharacters(in: .whitespacesAndNewlines).prefix(limit))
  }

  private struct PathAccess {
    var path: String
    var mode: AgentNativeResourceLockMode
  }

  private static let globalKey = "global"
  private static let maxIdentifierCharacters = 512
  private static let maxPathComponents = 256
  private static let maxPathComponentCharacters = 512
  private static let exclusiveGlobalMutationToolIds: Set<String> = [
    AgentIOSProjectRepositoryMutationToolCatalog.commit,
    AgentIOSProjectRepositoryMutationToolCatalog.push
  ]
}

enum AgentNativeToolResourceLockTable {
  private static let tableLock = NSLock()
  private static var entries: [String: Entry] = [:]

  static func execute<T>(
    plan: AgentNativeResourceLockPlan,
    checkpoint: () throws -> Void,
    operation: () throws -> T
  ) rethrows -> T {
    let reservations = reserve(plan.requests)
    var acquired: [EntryLock] = []
    defer {
      acquired.reversed().forEach { $0.release() }
      release(reservations)
    }
    for reservation in reservations {
      try reservation.entry.lock.acquire(mode: reservation.request.mode, checkpoint: checkpoint)
      acquired.append(reservation.entry.lock)
    }
    try checkpoint()
    return try operation()
  }

  private static func reserve(_ requests: [AgentNativeResourceLockRequest]) -> [Reservation] {
    tableLock.lock()
    defer { tableLock.unlock() }
    return requests.map { request in
      let entry = entries[request.key] ?? Entry()
      entries[request.key] = entry
      entry.references += 1
      return Reservation(request: request, entry: entry)
    }
  }

  private static func release(_ reservations: [Reservation]) {
    tableLock.lock()
    defer { tableLock.unlock() }
    for reservation in reservations {
      reservation.entry.references -= 1
      precondition(reservation.entry.references >= 0, "Resource lock reference count underflow")
      if reservation.entry.references == 0, entries[reservation.request.key] === reservation.entry {
        entries.removeValue(forKey: reservation.request.key)
      }
    }
  }

  private final class Entry {
    let lock = EntryLock()
    var references = 0
  }

  private final class EntryLock {
    private let condition = NSCondition()
    private var readers = 0
    private var writerActive = false
    private var waitingWriters = 0

    func acquire(mode: AgentNativeResourceLockMode, checkpoint: () throws -> Void) rethrows {
      condition.lock()
      if mode == .write { waitingWriters += 1 }
      defer {
        if mode == .write { waitingWriters = max(waitingWriters - 1, 0) }
        condition.unlock()
      }
      while mode == .read
        ? (writerActive || waitingWriters > 0)
        : (writerActive || readers > 0) {
        _ = condition.wait(until: Date().addingTimeInterval(0.1))
        condition.unlock()
        defer { condition.lock() }
        try checkpoint()
      }
      if mode == .read { readers += 1 } else { writerActive = true }
    }

    func release() {
      condition.lock()
      if writerActive {
        writerActive = false
      } else {
        precondition(readers > 0, "Resource read lock released without an owner")
        readers -= 1
      }
      condition.broadcast()
      condition.unlock()
    }
  }

  private struct Reservation {
    var request: AgentNativeResourceLockRequest
    var entry: Entry
  }
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/AgentNativeToolResourceLocks.kt
package com.galaxyssi.chat

import java.security.MessageDigest
import java.util.Locale
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantReadWriteLock
import org.json.JSONArray
import org.json.JSONObject

internal enum class AgentNativeResourceLockMode {
    READ,
    WRITE
}

internal data class AgentNativeResourceLockRequest(
    val key: String,
    val mode: AgentNativeResourceLockMode
)

internal data class AgentNativeResourceLockPlan(
    val requests: List<AgentNativeResourceLockRequest>,
    val resourceScoped: Boolean
) {
    fun conflictsWith(other: AgentNativeResourceLockPlan): Boolean {
        val right = other.requests.associateBy(AgentNativeResourceLockRequest::key)
        return requests.any { left ->
            val matching = right[left.key] ?: return@any false
            left.mode == AgentNativeResourceLockMode.WRITE ||
                matching.mode == AgentNativeResourceLockMode.WRITE
        }
    }
}

/** Resolves workspace and hierarchical path locks without exposing local path names. */
internal object AgentNativeToolResourcePolicy {
    fun resolve(
        descriptor: AgentNativeToolDescriptor,
        input: AgentNativeJsonObject,
        fallbackWorkspaceId: String = ""
    ): AgentNativeResourceLockPlan {
        val operationMode = if (
            descriptor.concurrency == AgentNativeToolConcurrency.PARALLEL_READ_ONLY
        ) {
            AgentNativeResourceLockMode.READ
        } else {
            AgentNativeResourceLockMode.WRITE
        }
        if (input.requiresExclusiveVerification() || descriptor.requiresExclusiveGlobalMutation()) {
            return globalPlan(AgentNativeResourceLockMode.WRITE)
        }
        if (!descriptor.supportsWorkspaceScoping()) {
            return globalPlan(operationMode)
        }
        val workspaceId = descriptor.workspaceIdentity(input, fallbackWorkspaceId)
        if (workspaceId.isBlank()) return globalPlan(operationMode)

        val workspaceKey = "workspace:${stableSegment(workspaceId)}"
        val pathAccesses = if (descriptor.capabilities.any { "workspace" in it }) {
            collectPathAccesses(descriptor, input, operationMode)
        } else {
            emptyList()
        }
        val requests = mutableListOf(
            AgentNativeResourceLockRequest(GLOBAL_KEY, AgentNativeResourceLockMode.READ)
        )
        if (pathAccesses.isEmpty()) {
            requests += AgentNativeResourceLockRequest(workspaceKey, operationMode)
        } else {
            requests += AgentNativeResourceLockRequest(workspaceKey, AgentNativeResourceLockMode.READ)
            pathAccesses.forEach { access ->
                val components = normalizePath(access.path)
                if (components.isEmpty()) {
                    requests += AgentNativeResourceLockRequest(workspaceKey, access.mode)
                } else {
                    components.indices.forEach { index ->
                        val prefix = components.take(index + 1).joinToString("/")
                        requests += AgentNativeResourceLockRequest(
                            key = "path:${stableSegment(workspaceId)}:${stableSegment(prefix)}",
                            mode = if (index == components.lastIndex) {
                                access.mode
                            } else {
                                AgentNativeResourceLockMode.READ
                            }
                        )
                    }
                }
            }
        }
        return AgentNativeResourceLockPlan(normalizeRequests(requests), resourceScoped = true)
    }

    fun resolveAction(
        descriptor: AgentNativeToolDescriptor,
        action: AgentAction,
        fallbackWorkspaceId: String = ""
    ): AgentNativeResourceLockPlan? {
        val input = parseInput(action.parameters["input_json"].orEmpty()) ?: return null
        return resolve(descriptor, input, fallbackWorkspaceId)
    }

    private fun globalPlan(mode: AgentNativeResourceLockMode) = AgentNativeResourceLockPlan(
        requests = listOf(AgentNativeResourceLockRequest(GLOBAL_KEY, mode)),
        resourceScoped = false
    )

    private fun AgentNativeToolDescriptor.supportsWorkspaceScoping(): Boolean =
        capabilities.any { capability ->
            "workspace" in capability || capability.startsWith("runtime.") ||
                capability.startsWith("project.")
        }

    private fun AgentNativeToolDescriptor.requiresExclusiveGlobalMutation(): Boolean =
        id in EXCLUSIVE_GLOBAL_MUTATION_TOOL_IDS

    private val EXCLUSIVE_GLOBAL_MUTATION_TOOL_IDS = setOf(
        AgentMobileProjectNativeTools.COMMIT,
        AgentMobileProjectNativeTools.PUSH,
        AgentMobileProjectNativeTools.CREATE_PULL_REQUEST,
        AgentMobileProjectNativeTools.PUBLISH_PULL_REQUEST,
        AgentMobileProjectNativeTools.FINALIZE_PULL_REQUEST
    )

    private fun AgentNativeToolDescriptor.workspaceIdentity(
        input: AgentNativeJsonObject,
        fallbackWorkspaceId: String
    ): String {
        val explicit = input.stringValue("workspace_id")
        if (explicit.isNotBlank()) return "${location.wireValue}:$explicit"
        val desktopId = input.stringValue("desktop_id")
        if (desktopId.isNotBlank()) {
            return "${location.wireValue}:$desktopId:${fallbackWorkspaceId.ifBlank { "default" }}"
        }
        return fallbackWorkspaceId.trim().take(MAX_ID_CHARS)
            .takeIf(String::isNotBlank)
            ?.let { "${location.wireValue}:$it" }
            .orEmpty()
    }

    private fun collectPathAccesses(
        descriptor: AgentNativeToolDescriptor,
        input: AgentNativeJsonObject,
        operationMode: AgentNativeResourceLockMode
    ): List<PathAccess> = buildList {
        fun visit(key: String, value: Any?) {
            when (value) {
                is String -> if (key.isPathKey()) {
                    add(PathAccess(value, pathMode(descriptor.id, key, input, operationMode)))
                }
                is Map<*, *> -> value.forEach { (childKey, childValue) ->
                    if (childKey is String) visit(childKey, childValue)
                }
                is Iterable<*> -> value.forEach { child ->
                    if (key.isPathCollectionKey() && child is String) {
                        add(PathAccess(child, pathMode(descriptor.id, key, input, operationMode)))
                    } else {
                        visit(key, child)
                    }
                }
                is Array<*> -> value.forEach { child -> visit(key, child) }
            }
        }
        input.forEach { (key, value) -> visit(key, value) }
    }.filter { it.path.isNotBlank() }

    private fun pathMode(
        toolId: String,
        key: String,
        input: AgentNativeJsonObject,
        operationMode: AgentNativeResourceLockMode
    ): AgentNativeResourceLockMode {
        if (operationMode == AgentNativeResourceLockMode.READ) return operationMode
        val normalizedKey = key.lowercase(Locale.ROOT)
        if (normalizedKey == "source_path" || normalizedKey == "source_paths") {
            return if (toolId.endsWith(".move")) operationMode else AgentNativeResourceLockMode.READ
        }
        if (normalizedKey == "paths") return AgentNativeResourceLockMode.READ
        if (normalizedKey == "archive_path" &&
            (toolId.endsWith(".zip.extract") || toolId.endsWith(".zip.list"))
        ) {
            return AgentNativeResourceLockMode.READ
        }
        if (normalizedKey == "path" && input.containsKey("output_path")) {
            return AgentNativeResourceLockMode.READ
        }
        return operationMode
    }

    private fun String.isPathKey(): Boolean {
        val normalized = lowercase(Locale.ROOT)
        return normalized == "path" || normalized.endsWith("_path")
    }

    private fun String.isPathCollectionKey(): Boolean {
        val normalized = lowercase(Locale.ROOT)
        return normalized == "paths" || normalized.endsWith("_paths")
    }

    private fun normalizePath(value: String): List<String> {
        val normalized = mutableListOf<String>()
        value.replace('\\', '/').split('/').forEach { component ->
            when (component.trim()) {
                "", "." -> Unit
                ".." -> if (normalized.isNotEmpty()) normalized.removeAt(normalized.lastIndex)
                else -> normalized += component.trim().take(MAX_PATH_COMPONENT_CHARS)
            }
        }
        return normalized.take(MAX_PATH_COMPONENTS)
    }

    private fun normalizeRequests(
        requests: List<AgentNativeResourceLockRequest>
    ): List<AgentNativeResourceLockRequest> = requests
        .groupBy(AgentNativeResourceLockRequest::key)
        .map { (key, grouped) ->
            AgentNativeResourceLockRequest(
                key,
                if (grouped.any { it.mode == AgentNativeResourceLockMode.WRITE }) {
                    AgentNativeResourceLockMode.WRITE
                } else {
                    AgentNativeResourceLockMode.READ
                }
            )
        }
        .sortedBy(AgentNativeResourceLockRequest::key)

    private fun parseInput(raw: String): AgentNativeJsonObject? = runCatching {
        JSONObject(raw.ifBlank { "{}" }).toNativeJsonObject()
    }.getOrNull()

    private fun JSONObject.toNativeJsonObject(): AgentNativeJsonObject = keys().asSequence()
        .associateWith { key -> get(key).toNativeJsonValue() }

    private fun Any?.toNativeJsonValue(): Any? = when (this) {
        JSONObject.NULL -> null
        is JSONObject -> toNativeJsonObject()
        is JSONArray -> (0 until length()).map { index -> get(index).toNativeJsonValue() }
        else -> this
    }

    private fun AgentNativeJsonObject.stringValue(key: String): String =
        (this[key] as? String).orEmpty().trim().take(MAX_ID_CHARS)

    private fun AgentNativeJsonObject.requiresExclusiveVerification(): Boolean =
        stringValue("verification_kind")
            .lowercase(Locale.ROOT)
            .let { it.isNotBlank() && it != "none" }

    private fun stableSegment(value: String): String = MessageDigest.getInstance("SHA-256")
        .digest(value.toByteArray(Charsets.UTF_8))
        .take(16)
        .joinToString("") { byte -> "%02x".format(Locale.ROOT, byte.toInt() and 0xff) }

    private data class PathAccess(
        val path: String,
        val mode: AgentNativeResourceLockMode
    )

    private const val GLOBAL_KEY = "global"
    private const val MAX_ID_CHARS = 512
    private const val MAX_PATH_COMPONENTS = 256
    private const val MAX_PATH_COMPONENT_CHARS = 512
}

/** Fair, reference-counted lock table shared by all native-tool registries. */
internal object AgentNativeToolResourceLockTable {
    private val tableGuard = Any()
    private val entries = mutableMapOf<String, Entry>()

    fun <T> execute(
        plan: AgentNativeResourceLockPlan,
        checkpoint: () -> Unit,
        block: () -> T
    ): T {
        val reservations = reserve(plan.requests)
        val acquired = mutableListOf<Lock>()
        try {
            reservations.forEach { reservation ->
                val lock = when (reservation.request.mode) {
                    AgentNativeResourceLockMode.READ -> reservation.entry.lock.readLock()
                    AgentNativeResourceLockMode.WRITE -> reservation.entry.lock.writeLock()
                }
                acquire(lock, checkpoint)
                acquired += lock
            }
            checkpoint()
            return block()
        } finally {
            acquired.asReversed().forEach(Lock::unlock)
            release(reservations)
        }
    }

    private fun reserve(requests: List<AgentNativeResourceLockRequest>): List<Reservation> =
        synchronized(tableGuard) {
            requests.map { request ->
                val entry = entries.getOrPut(request.key) { Entry() }
                entry.references += 1
                Reservation(request, entry)
            }
        }

    private fun release(reservations: List<Reservation>) = synchronized(tableGuard) {
        reservations.forEach { reservation ->
            reservation.entry.references -= 1
            check(reservation.entry.references >= 0) { "Resource lock reference count underflow" }
            if (reservation.entry.references == 0) {
                entries.remove(reservation.request.key, reservation.entry)
            }
        }
    }

    private fun acquire(lock: Lock, checkpoint: () -> Unit) {
        while (true) {
            checkpoint()
            val acquired = try {
                lock.tryLock(POLL_MILLIS, TimeUnit.MILLISECONDS)
            } catch (_: InterruptedException) {
                Thread.currentThread().interrupt()
                throw AgentNativeToolCancelledException()
            }
            if (acquired) return
        }
    }

    private class Entry(
        val lock: ReentrantReadWriteLock = ReentrantReadWriteLock(true),
        var references: Int = 0
    )

    private data class Reservation(
        val request: AgentNativeResourceLockRequest,
        val entry: Entry
    )

    private const val POLL_MILLIS = 100L
}
